### PR TITLE
Add RFC to remove non-standard Github teams

### PR DESCRIPTION
**Pull request by Gerg**
🌶

https://cloudfoundry.slack.com/archives/C026XJGNC2U/p1651186805030849

### CODE REVIEW

**Review by ameowlia — APPROVED** [review]

**Comment by rkoster** [comment]
Before this can be made a reality we will need to start tracking ci/bot accounts used by the different working groups in their charters.

In addition, how do we want to deal with repos containing ci related configuration for a ci instance maintained by a single vendor? For those repos it makes sense to give the whole team of the vendor access, maybe these repos should not be part of the cloudfoundry org, or ci should be moved over to common infrastructure. However, both options require effort, and probably should not block the adoption of this process.

A pragmatic solution would be to generate `org/cloudfoundry.yml` based on the working group charters, and then merge in a directory with yml files containing exceptions. This would clearly document the exceptions, which allows us to define processes for these down the line.

**Comment by stephanme** [comment]
My proposal would be to first generate the new teams conforming to [rfc-0005-github-teams-and-access](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0005-github-teams-and-access.md) in addition to the existing teams. This allows all WGs to start deleting unneeded teams (because they got replaced by the new ones) or at least remove the majority of members in the non-standard teams.

Once #248 is merged, I can add the team generation. Need to find a time slot with @rkoster .

**Comment by stephanme** [comment]
Two more PRs to finaliye the automatic github team generation:
#316 - Generate github teams for toc and wg-leads
#317 - Enable generation of github teams for all WGs

**Comment by davewalter** [comment]
Could someone please post a list of the teams that are scheduled for removal?

**Comment by Gerg** [comment]
This change will likely hit private repositories hardest, since they don't appear to be linked to working groups.

**Comment by emalm** [comment]
@davewalter This proposal would remove all the entries under `orgs.cloudfoundry.teams` in https://github.com/cloudfoundry/community/blob/main/org/cloudfoundry.yml:  https://github.com/cloudfoundry/community/blob/05f17542a044fb114e38b26808e70d2ed0633214/org/cloudfoundry.yml#L2283-L6650

The automation process now generates the teams specified in [RFC 0005](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0005-github-teams-and-access.md) from the YAML embedded in the WG charter definitions and applies them to the CF GitHub org.

@Gerg and other @cloudfoundry/wg-leads, if there are existing private repos that WG areas should continue to have access to, please propose adding them to the appropriate charters.

@cloudfoundry/toc I had forgotten when we were discussing during the TOC call today that this proposal was structured as an RFC that still needs ratification. If @Gerg approves as well, I suggest we vote asynchronously via comment to start the week-long Final Comment Period for this RFC, with the expectation of accepting the RFC next week. At that point, we will be cleared to carry out the team changes proposed here.

**Comment by Gerg** [comment]
Looks like the majority of private repos defined in `org/cloudfoundry.yml` are not assigned to working groups.

Total number of private repos:
```sh
$ yq '.orgs.cloudfoundry.repos | to_entries[] | select(.value.private != null) | .key' org/cloudfoundry.yml | wc -l
     106
```

Private repos appearing in working group charters (I'm manually omitting some false positives):
```sh
$ yq '.orgs.cloudfoundry.repos | to_entries[] | select(.value.private != null) | .key' org/cloudfoundry.yml | xargs -I _  rg _ ./toc/working-groups
./toc/working-groups/foundational-infrastructure.md
171:  - cloudfoundry/bosh-bbl-ci-envs
./toc/working-groups/foundational-infrastructure.md
176:  - cloudfoundry/bosh-cpi-environments
./toc/working-groups/foundational-infrastructure.md
178:  - cloudfoundry/bosh-cpi-kb
./toc/working-groups/foundational-infrastructure.md
200:  - cloudfoundry/bosh-workstation
./toc/working-groups/cf-on-k8s.md
67:  - cloudfoundry/cf-k8s-secrets
./toc/working-groups/app-runtime-platform.md
86:  - cloudfoundry/diego-team
...
./toc/working-groups/app-runtime-platform.md
145:  - cloudfoundry/garden-ci
146:  - cloudfoundry/garden-ci-artifacts-release
...
./toc/working-groups/app-runtime-platform.md
257:  - cloudfoundry/networking-oss-deployments
```

I suspect this is because our effort earlier this year to assign orphaned repositories excluded private repositories.

**Comment by stephanme** [comment]
This PR would only remove the non-standard teams listed in cloudfoundry.yml (orgs > cloudfoundry > teams). Individual contributors (and lots of other repository configuration options) are not listed in cloudfoundry.yml and not managed by peribolos.

https://github.com/relengfam/peribolos/blob/main/org/repos.go#L191-L205 gives an idea what is managed for a repo by the automation. Teams (members and repo access) are separate peribolos entities.

File: toc/rfc/rfc-draft-remove-nonstandard-github-teams.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+Github UI and intermediary tools like
+[`cloudfoundry.yml`](https://github.com/cloudfoundry/community/blob/main/org/cloudfoundry.yml).
+
+## Proposal
+
+Delete all Github teams in the `cloudfoundry` organization that do not conform
+to
+[rfc-0005-github-teams-and-access](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0005-github-teams-and-access.md)
+or another CFF-recognized entity (for example: the
+[`toc`](https://github.com/orgs/cloudfoundry/teams/toc) team).
```
Comment (emalm) [inline toc/rfc/rfc-draft-remove-nonstandard-github-teams.md:31]: * Point out CFF staff as a better example of a team outside of the CFF technical governance structure.
* Add a proviso for temporary exceptional teams to smooth over access issues during the transition.
* Add a proviso for CFF staff to establish teams to access non-project repositories.

```suggestion
Delete all GitHub teams in the `cloudfoundry` organization that do not conform
to
[rfc-0005-github-teams-and-access](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0005-github-teams-and-access.md)
or another CFF-recognized entity, such as CFF staff members.

To allow the CF community to function while the Working Groups adjust their scope and membership, the TOC will permit the formation of exceptional teams on a case-by-case basis. These exceptional teams must be identified as such clearly in their names with phrases such as "temporary" and "delete me", and these teams should be related to unofficial teams that were previously deleted. The TOC and Working Groups will review these exceptional teams regularly for deletion, with the goal of removing all exceptional teams by the end of 2022.

CFF staff may create teams as needed to administer repositories outside the scope of technical project operation. The names for any such teams should clearly indicate their relation to CFF staff responsibilities.
```

File: toc/rfc/rfc-draft-remove-nonstandard-github-teams.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+## Problem
+
+There are currently many Github teams in the `cloudfoundry` Github organization
+that do not correlate with a CFF-recognized entity (working group, TOC, etc).
+These teams grant access to CFF-governed repositories, but are not beholden to
+the CFF's roles and other processes for governing access. In addition, the
+volume of Github teams makes it more difficult to manage team access via the
+Github UI and intermediary tools like
+[`cloudfoundry.yml`](https://github.com/cloudfoundry/community/blob/main/org/cloudfoundry.yml).
+
```
Comment (emalm) [inline toc/rfc/rfc-draft-remove-nonstandard-github-teams.md:24]: * Acknowledge short-term access problems coming from the clean sweep of legacy teams.
* Point out some teams may be needed for access to repositories outside of CFF technical projects.
```suggestion

Nevertheless, controlling access to Cloud Foundry repositories only by teams defined by CFF technical governance is a substantial change from the previous system of access and team management and in the short term will cause unexpected loss of access to repositories for some contributors and automated development and release processes.

As per [RFC 0007](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0007-repository-ownership.md), CFF staff also require access to many repositories in the Cloud Foundry GitHub organization that are outside the scope of project technical operation, such as those for CFF websites or the Cloud Foundry Certified Developer examinations.
```

File: toc/rfc/rfc-draft-remove-nonstandard-github-teams.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+- Name: Remove Non-Standard Github Teams
+- Start Date: 2022-04-28
+- Author(s): @gerg
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: [https://github.com/cloudfoundry/community/pull/262](https://github.com/cloudfoundry/community/pull/262)
+
+
+## Summary
+
+Remove Github teams (and associated access) that are outside of the CFF
```
Comment (emalm) [inline toc/rfc/rfc-draft-remove-nonstandard-github-teams.md:12]:
```suggestion
Remove GitHub teams (and associated access) that are outside of the CFF
```

File: toc/rfc/rfc-draft-remove-nonstandard-github-teams.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+governance model.
+
+## Problem
+
+There are currently many Github teams in the `cloudfoundry` Github organization
```
Comment (emalm) [inline toc/rfc/rfc-draft-remove-nonstandard-github-teams.md:17]:
```suggestion
There are currently many GitHub teams in the `cloudfoundry` Github organization
```

File: toc/rfc/rfc-draft-remove-nonstandard-github-teams.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+that do not correlate with a CFF-recognized entity (working group, TOC, etc).
+These teams grant access to CFF-governed repositories, but are not beholden to
+the CFF's roles and other processes for governing access. In addition, the
+volume of Github teams makes it more difficult to manage team access via the
+Github UI and intermediary tools like
```
Comment (emalm) [inline toc/rfc/rfc-draft-remove-nonstandard-github-teams.md:22]:
```suggestion
volume of GitHub teams makes it more difficult to manage team access via the
GitHub UI and intermediary tools like
```

File: toc/rfc/rfc-draft-remove-nonstandard-github-teams.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+[`cloudfoundry.yml`](https://github.com/cloudfoundry/community/blob/main/org/cloudfoundry.yml).
+
+## Proposal
+
+Delete all Github teams in the `cloudfoundry` organization that do not conform
```
Comment (emalm) [inline toc/rfc/rfc-draft-remove-nonstandard-github-teams.md:27]:
```suggestion
Delete all GitHub teams in the `cloudfoundry` organization that do not conform
```

File: toc/rfc/rfc-draft-remove-nonstandard-github-teams.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# Meta
+[meta]: #meta
+- Name: Remove Non-Standard Github Teams
```
Comment (emalm) [inline toc/rfc/rfc-draft-remove-nonstandard-github-teams.md:3]:
```suggestion
- Name: Remove Non-Standard GitHub Teams
```

**Review by emalm — CHANGES_REQUESTED** [review]
Suggested some exemptions for CFF staff teams and for temporary teams to ease over the transition. Also, `s/Github/GitHub/g`.

**Comment by emalm** [comment]
@Gerg Unless you have any objections, I'll add my suggested changes to the RFC draft before the TOC meeting next week.

**Comment by Gerg** [comment]
@emalm Sorry, busy week and I lost track of this. I went ahead and incorporated your changes. Looks good!

**Review by emalm — APPROVED** [review]

**Comment by emalm** [comment]
@cloudfoundry/toc Let's make sure we discuss starting the Final Comment Period for this RFC at tomorrow's meeting.

**Review by stephanme — APPROVED** [review]

**Review by AP-Hunt — APPROVED** [review]

**Review by rkoster — APPROVED** [review]

**Pull request merged**